Closes #2277 - Add ability to set default search engine

ekager opened the pull request:
I think since the calls to SearchEngineManager `getDefaultSearchEngine` are tested in `SearchEngineManagerTest`, this PR does need its own tests? 

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

ekager commented:
I'm working on setting a default search engine to actually be used for searches in Fenix, and @sblatz is working on search suggestions which will need one-off searches outside of the default engine. Creating a method to set the default makes sense, but we'll also need a way for the overrides, this was aiming to do both but I can see how passing in the default every time will get annoying.

pocmo commented:
Ah, okay, that makes sense.

How does this sound:
* For the default we add this to the SearchEngineManager - so that everyone can use the default without needing to know the name in advance.
* For the use case we allow passing in an optional`SearchEngine` object. If provided the use case will use this instead of `getDefaultSearchEngine`.

ekager commented:
Yeah, that sounds good! I'll work on making those changes

ekager commented:
So I talked to @sblatz more about this PR, and he says because he'll be able to add and remove providers, we don't think we will need to have the ability for a one-off search with a specific search engine (at least for now!). I've removed the `SearchUseCases` code, and replaced it with just a way to set the default search engine in `SearchEngineManager` (with tests!)! Let me know if this is the right way to do this :)

pocmo commented on components/browser/search/src/main/java/mozilla/components/browser/search/SearchEngineManager.kt:
```diff
@@ -28,6 +28,7 @@ class SearchEngineManager(
 ) {
     private var deferredSearchEngines: Deferred<List<SearchEngine>>? = null
     private val scope = CoroutineScope(Dispatchers.Default)
+    var defaultSearchEngine: SearchEngine? = null
```
I think in `SearchEngineManager` it may be good to still use the name and only have a setter for it (e.g. `setDefault(name: String)`) Otherwise it may be confusing if you look at it from the outside and you have two way to get to a `SearchEngine` - calling `getDefaultSearchEngine` but also `defaultSearchEngine`. But the second one only returns something you have set on the manager manually and not really the default in all cases. :)

pocmo commented on components/browser/search/src/main/java/mozilla/components/browser/search/SearchEngineManager.kt:
```diff
@@ -66,11 +67,21 @@ class SearchEngineManager(
+        val searchEngines = getSearchEngines(context)
+
+        return searchEngines.find { it.name == name }
+    }
```
Will this be needed after we change the code to allow setting the name directly?

ekager replied:
I think this will still be needed in order for @sblatz to pass a non-default search engine to the invoke function?

pocmo commented on components/feature/search/src/main/java/mozilla/components/feature/search/SearchUseCases.kt:
```diff
@@ -37,8 +38,8 @@ class SearchUseCases(
-            invoke(searchTerms, sessionManager.selectedSession)
+        override fun invoke(searchTerms: String, searchEngine: SearchEngine?) {
+            invoke(searchTerms, sessionManager.selectedSession, searchEngine)
         }
```
I wonder if we should add the `searchEngine` parameter to the constructor instead of the `invoke` method: Other components will call `invoke` and not provide the parameter (because they have no knowledge about the default). So instead the app needs to configure the use case to use the right search engine when something calls `invoke` on it.

ekager commented:
Okay, I created the setDefaultSearchEngine function that takes the engineName as a String, and removed the getEngineByName function. Now TaskCluster is complaining because IntentProcessor uses invoke but does not know about SearchEngine...? If I import it, it's an unused import, do you know how to fix this?

/build/android-components/components/feature/intent/src/main/java/mozilla/components/feature/intent/IntentProcessor.kt: (82, 49): Cannot access class 'mozilla.components.browser.search.SearchEngine'. Check your module classpath for missing or conflicting dependencies
